internal/cloud/realm/client.go

```diff
@@ -91,5 +91,32 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 		req.Header.Set(api.HeaderAuthorization, "Bearer "+auth)
 	}
 	client := &http.Client{}
-	return client.Do(req)
+	res, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode >= 200 && res.StatusCode <= 299 {
+		return res, nil
+	}
+
+	parsedErr := parseResponseError(res)
+	if err, ok := parsedErr.(ServerError); !ok {
+		return nil, parsedErr
+	} else if err.Code != invalidSessionCode {
+		return nil, parsedErr
+	}
+
+	if options.PreventRefresh {
+		return nil, parsedErr
+	}
+
+	authToken, refreshErr := c.refreshAuth()
+	if refreshErr != nil {
+		return nil, refreshErr
```

Comment:
`return nil, ErrInvalidSession` here.  if _anything_ fails in the `refreshAuth` method, we'll chalk it up to "unable to refresh your access token" and clear out your session